public/app/features/correlations/CorrelationsPage.tsx

```diff
@@ -0,0 +1,215 @@
+
+  return (
+    <Page navModel={navModel}>
+      <Page.Contents>
+        <div>
+          <HorizontalGroup justify="space-between">
+            <div>
+              <h4>Correlations</h4>
+              <p>Define how data living in different data sources relates to each other.</p>
+            </div>
+            {canWriteCorrelations && data?.length !== 0 && data !== undefined && !isAdding && (
+              <Button icon="plus" onClick={() => setIsAdding(true)}>
+                Add new
+              </Button>
+            )}
+          </HorizontalGroup>
+        </div>
+
+        {!data && get.loading && (
+          <div className={loaderWrapper}>
+            <LoadingPlaceholder text="loading..." />
+          </div>
+        )}
+
+        {showEmptyListCTA && <EmptyCorrelationsCTA onClick={() => setIsAdding(true)} />}
+
+        {
+          // This error is not actionable, it'd be nice to have a recovery button
+          get.error && get.error.status !== 404 && (
+            <Alert severity="error" title="Error fetching correlation data" topSpacing={2}>
+              <HorizontalGroup>
+                {get.error.data.message ||
+                  'An unknown error occurred while fetching correlation data. Please try again.'}
+              </HorizontalGroup>
+            </Alert>
+          )
+        }
+
+        {isAdding && <AddCorrelationForm onClose={() => setIsAdding(false)} onCreated={handleAdd} />}
+
+        {data && data.length >= 1 && (
+          <Table
+            renderExpandedRow={({ target, source, ...correlation }) => (
+              <EditCorrelationForm
+                defaultValues={{ sourceUID: source.uid, ...correlation }}
+                onUpdated={handleUpdate}
+                readOnly={isSourceReadOnly({ source }) || !canWriteCorrelations}
+              />
+            )}
+            columns={columns}
+            data={data}
+            getRowId={(correlation) => `${correlation.source.uid}-${correlation.uid}`}
+          />
+        )}
+      </Page.Contents>
+    </Page>
+  );
+}
+
+const getDatasourceCellStyles = (theme: GrafanaTheme2) => ({
+  root: css`
+    display: flex;
+    align-items: center;
+  `,
+  dsLogo: css`
+    margin-right: ${theme.spacing()};
+    height: 16px;
+    width: 16px;
+  `,
+});
+
+const DataSourceCell = memo(
+  function DataSourceCell({
+    cell: { value },
+  }: CellProps<CorrelationData, CorrelationData['source'] | CorrelationData['target']>) {
+    const styles = useStyles2(getDatasourceCellStyles);
+
+    return (
+      <span className={styles.root}>
+        <img src={value.meta.info.logos.small} className={styles.dsLogo} />
+        {value.name}
+      </span>
+    );
+  },
+  ({ cell: { value } }, { cell: { value: prevValue } }) => {
+    return value.type === prevValue.type && value.name === prevValue.name;
+  }
+);
+
+const noWrap = css`
+  white-space: nowrap;
+`;
```

Comment:
In order to have all styling in one place I'd suggest to add this below line 179.

Reply:
i do not agree. there are 3 styles objects/functions in this file, each one is defined just before the component using them.